app/controllers/topics_controller.rb

```diff
@@ -1314,13 +1313,27 @@ def track_visit_to_topic
       TopicsController.defer_add_incoming_link(hash)
     end
 
-    TopicsController.defer_track_visit(topic_id, ip, user_id, track_visit)
+    TopicsController.defer_track_visit(topic_id, user_id) if should_track_visit_to_topic?
   end
 
-  def self.defer_track_visit(topic_id, ip, user_id, track_visit)
+  def self.defer_track_visit(topic_id, user_id)
     Scheduler::Defer.later "Track Visit" do
+      TopicUser.track_visit!(topic_id, user_id)
+    end
+  end
+
+  def self.defer_topic_view(topic_id, ip, user_id = nil)
+    Scheduler::Defer.later "Topic View" do
+      topic = Topic.find_by(id: topic_id)
```

Comment:
Kind of annoying we now have to do some AR lookups here, but we got the security checking for free in the old TopicsController#show route

Reply:
hmmm but now we get security here but do not get it for tracking topic views? so users can potentially log views on topics they have not permission to see? can't we centralize all the deferred work?